kubernetes/config/incluster_config.c

```diff
@@ -0,0 +1,154 @@
+#include "incluster_config.h"
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#define SERVICE_HOST_ENV_NAME "KUBERNETES_SERVICE_HOST"
+#define SERVICE_PORT_ENV_NAME "KUBERNETES_SERVICE_PORT"
+#define SERVICE_HTTPS_PREFIX "https://"
+#define SERVICE_TOKEN_FILENAME "/var/run/secrets/kubernetes.io/serviceaccount/token"
+#define SERVICE_CERT_FILENAME "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+#define AUTH_TOKEN_KEY "Authorization"
+#define BEARER_TOKEN_TEMPLATE "Bearer %s"
+#define BEARER_TOKEN_BUFFER_SIZE 1024
+
+static int checkServiceAccountFile(const char *fileName)
+{
+    static char fname[] = "checkServiceAccountFile()";
+
+    if (!fileName) {
+        fprintf(stderr, "%s: The file name must be specified.\n", fname);
+        return -1;
+    }
+
+    if (-1 == access(fileName, F_OK)) {
+        fprintf(stderr, "%s: The file %s does not exist.[%s]\n", fname, fileName, strerror(errno));
+        return -1;
+    }
+
+    if (-1 == access(fileName, R_OK)) {
+        fprintf(stderr, "%s: No permission to read the file %s.[%s]\n", fname, fileName, strerror(errno));
+        return -1;
+    }
+
+    return 0;
+}
+
+static int setBasePathInCluster(char **pBasePath)
+{
+    static char fname[] = "setBasePathInCluster()";
+
+    const char *service_host_env = getenv(SERVICE_HOST_ENV_NAME);
+    if (!service_host_env) {
+        fprintf(stderr, "%s: Cannot retrieve the kubernetes service host inside a pod by the env %s.\n", fname, SERVICE_HOST_ENV_NAME);
+        return -1;
+    }
+
+    const char *service_port_env = getenv(SERVICE_PORT_ENV_NAME);
+    if (!service_port_env) {
+        fprintf(stderr, "%s: Cannot retrieve the kubernetes service port inside a pod by the env %s.\n", fname, SERVICE_PORT_ENV_NAME);
+        return -1;
+    }
+
+    char *basePath = calloc(strlen(SERVICE_HTTPS_PREFIX) + strlen(service_host_env) + strlen(service_port_env) + 2 /* 1 for ':', 1 for '\0' */ , sizeof(char));
+    if (!basePath) {
+        fprintf(stderr, "%s: Cannot allocate the memory for base path for kubernetes service.\n", fname);
+        return -1;
+    }
+
+    sprintf(basePath, "%s%s:%s", SERVICE_HTTPS_PREFIX, service_host_env, service_port_env);
+    *pBasePath = basePath;
+    return 0;
+}
+
+static int setSslConfigInCluster(sslConfig_t ** pSslConfig, const char *caFileName)
+{
+    const char *ca_file = NULL;
+    int insecure_skip_tls_verify = 0;
+
+    if (0 == checkServiceAccountFile(caFileName)) {
+        ca_file = caFileName;
+    } else {
+        insecure_skip_tls_verify = 1;
```

Comment:
I don't think this is correct. In theory you could have installed the CA in the known-CAs path inside the Pod. Just because the file isn't there, doesn't mean you want to skip certificate validation.

Reply:
OK. I will change it.